### ISSUE TITLE
hclust algorithm implementation change in 0.1.2?

### ISSUE THREAD

**Issue by buddha2490** [kassambara/ggcorrplot#14]
I recently updated the ggcorrplot package to 0.1.2.  Running ggcorrplot() on a correlation matrix with the hc.order=T option, the ordering of the heatmap drastically changed compared to the older version.  I don't know if this is a bug or intentional, but I thought I should point it out to you.

**Comment by MichelineCampbell** [comment]
I am having the same problem with version 0.1.4. With hc.order = F, a correlation pair of r = -0.27 and p = 0.0549 is excluded (as it should be), but with hc,order = T, this correlation pair is included.

**Comment by kassambara** [comment]
Fixed in the dev version. When `hc.order = TRUE`, the clustering is now computed on the **unrounded** correlation matrix, so the internal rounding to `digits` no longer changes the ordering.

Here is a matrix where the rounding shifts the dendrogram — the order is now stable and follows the full-precision clustering:

```r
remotes::install_github("kassambara/ggcorrplot")
library(ggcorrplot)

m <- structure(c(1,0.6063,0.1987,0.2749,-0.2482,0.231, 0.6063,1,-0.0788,0.2389,-0.0252,0.4772,
  0.1987,-0.0788,1,0.7596,0.3316,-0.445, 0.2749,0.2389,0.7596,1,0.8052,0.5687,
  -0.2482,-0.0252,0.3316,0.8052,1,-0.144, 0.231,0.4772,-0.445,0.5687,-0.144,1),
  dim = c(6L, 6L), dimnames = list(LETTERS[1:6], LETTERS[1:6]))

ggcorrplot(m, hc.order = TRUE, lab = TRUE, digits = 1)
```
![](https://i.imgur.com/f5TDZAG.png)

It now orders by clustering on the full-precision matrix (`C D E F A B` here) rather than on the rounded values. Thanks for reporting this.

**Comment by buddha2490** [comment]
Oh my gosh, didn’t I point this out in like 2019?  There are a lot of people using your R package for their analyses, and if you are going to continue to make it available, you owe all of us some more timely updates.  There are hundreds of publications that have used this code, including mine.  If you are going to put an R package on CRAN, you need to address these issues in a timely manner, not once a decade.

**Comment by buddha2490** [comment]
Sorry, 2018!

**Comment by kassambara** [comment]
You're right that this took far too long, and I'm sorry for the delay — thank you for pushing on it.

For your reassurance and anyone else's: this only affects the *ordering* of the variables in the plot when `hc.order = TRUE` (the hierarchical-clustering rearrangement). It never changed the correlation coefficients, the p-values, or any statistical result — only how the tiles are arranged. So published correlations and conclusions are unaffected; at most a figure's variable order could have looked different.

The cause was that the matrix was rounded before clustering, so rounding could create ties that shifted the dendrogram. The dev version now clusters on the unrounded matrix, so the ordering is stable:

```r
remotes::install_github("kassambara/ggcorrplot")
```

Appreciate you flagging it.

**Comment by buddha2490** [comment]
This wasn’t a small deal in our publication, this changed the results.  You rounded numbers BEFORE they were used in subsequent analyses.  That completely changed the analyses and results we were working on.  The solution was that I cloned your repo, fixed your code, and reran the analyses we were working on.  But other users don’t know that you are transforming numbers for subsequent transformations.  That’s why it was bad.  I fixed it, I published the paper, but this cost me quite a long time in debugging.  That’s fine, I published the paper and everything worked out.  But you should not take 8 years to respond to a bug fix.  Bugs happen, but you need to take better responsibility for your CRAN-hosted code.

**Comment by buddha2490** [comment]
I want to be really clear.  I have given talks at major data science conferences, Posit Conf, UseR!, Pharma conferences.  I use your package as an example about why one needs to care about source control and really monitor open source packages.  I’ve been doing it for years and you are just now getting around to it.  I’ve used your package as a bad example of open source for 8 years.  You did significant damage  to the open source community by ignoring my bug report for EIGHT YEARS.

**Comment by kassambara** [comment]
Understood, it's fixed in the dev version now, and I'll try to keep turnaround tighter going forward. If you hit anything concrete that still looks off, open a fresh issue with a reprex and I'll take a look.

**Comment by buddha2490** [comment]
Do you understand that it is too late?   I'm off in another career right now, I don't really care about this package, I don't use it.  But you did real damage.  I referenced your package by name in public.  Your package is actually good, that's why people use it.  It is used in life sciences context that is subsequently use to affect people's lives.  Don't mess around with this. If you aren't going to make a commitment to maintain this properly, you need to remove it from CRAN.
